benchmarks/test_stress_single_workflow_multiple_runs.py

```diff
@@ -0,0 +1,208 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2021 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Benchmark - run single workflow multiple times, record results, produce report"""
+
+import json
+import logging
+import os
+import subprocess
+import time
+from argparse import ArgumentParser
+from datetime import datetime
+from pathlib import Path
+from typing import Dict
+
+import matplotlib.pyplot as plt
+import pandas as pd
+
+logging.basicConfig(format="%(asctime)s %(message)s", level=logging.INFO)
+
+
+def run_workflows(workflow_name: str, n: int):
+    logging.info("Running workflow multiple times...")
+    for _ in range(0, n):
+        subprocess.run(
+            ["reana-client", "run", "-w", workflow_name, "--skip-validation"],
+            stdout=subprocess.DEVNULL,
+        )
+
+
+def get_results(workflow_name: str) -> pd.DataFrame:
+    return pd.DataFrame(
+        json.loads(
+            subprocess.check_output(
+                ["reana-client", "list", "--json", "--filter", f"name={workflow_name}"]
+            ).decode("ascii")
+        )
+    )
+
+
+def get_results_when_completed(workflow_name) -> pd.DataFrame:
+    logging.info("Waiting for runs to complete...")
```

Comment:
Whilst the script waits for results, it might be good to monitor the memory usage and the number of DB connections.

Here is an example whilst running 1000 hello world workflows:

```console
$ kubectl top nodes
NAME                 CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
kind-control-plane   3743m        31%    11163Mi         13%  

$ kubectl exec -i -t deployment/reana-db -- psql -U reana -c 'SELECT COUNT(*) FROM pg_stat_activity'   
 count 
-------
    53
(1 row)

$ kubectl get pods | grep -v STATUS | awk '{print $3}' | sort|uniq -c|sort -rn|head -20|awk '!max{max=$1;}{r="";i=s=60*$1/max;while(i-->0)r=r"#"; printf "%15s %5d %s %s",$2,$1,r,"\n";}'
        Pending    28 ############################################################ 
    Terminating    26 ######################################################## 
        Running    17 ##################################### 
ContainerCreating     4 ######### 
```


For example, the script could run this either once, or every 30 seconds, or so, and output results to the log file.

(Currently I'm doing it manually in another terminal...)

Reply:
Ditto for workflow status, not only pods status, for example:

```console
$ reana-client list --filter name=hello-serial-1000 | awk '{print $6}' | grep -v STATUS |  sort|uniq -c|sort -rn|head -20|awk '!max{max=$1;}{r="";i=s=60*$1/max;while(i-->0)r=r"#"; printf "%15s %5d %s %s",$2,$1,r,"\n";}'
         queued   535 ############################################################ 
       finished   436 ################################################# 
        running    21 ### 
        pending     5 # 
        created     1 # 
```

where `hello-serial-1000` is what I used as `-w` when running the benchmark.

Reply:
It would be good but I would prefer to keep the script simple (it is already getting quite big). This script is responsible for sequentially running the same workflows multiple times. We can create a couple of other scripts that will monitor, log performance or run workflows differently. Afterwards, we can connect some of them into the more sophisticated analysis like "run multiple workflows, at the same time monitor their pods, at the same time monitor amount of the DB connections they make".